### PR TITLE
Allow dynamic topography calculation from surface density contrast

### PR DESCRIPTION
**Pull request by Shangxin-Liu**
Sometimes we want to compute dynamic topography underneath ocean or other surface geological body. This needs the density contrast throughout the top boundary instead of only density value of the top cell. I set the default value of outside density to zero.

### CODE REVIEW

File: source/postprocess/dynamic_topography.cc [hunk 1]
```diff
@@ -257,7 +257,9 @@ namespace aspect
                              "in the outputted data file (not visualization). "
                              "'true' subtracts the mean, 'false' leaves "
                              "the calculated dynamic topography as is. ");
-
+          prm.declare_entry ("Density above","0",
+                             Patterns::Double (0),
+                             "The density value out of the surface boundary.");
```
Comment (bangerth) [inline source/postprocess/dynamic_topography.cc]: I think this is too short an explanation. Can you explain here and in the other place in more detail what this parameter does?

Comment (Shangxin-Liu) [inline source/postprocess/dynamic_topography.cc]: Sure. Here are the new ones with detailed parameter explanation. I've talked with Rene @gassmoeller  last week on this. The current ASPECT geodynamic topo calculation defaults the outside density to 0 (air). But actually the equation should use the surface density contrast, especially when we want to compute it underneath ocean it needs to be sea water density. More detail can be found here:  https://geodynamics.org/cig/files/8914/1158/8914/CIG_Poster_dyntopo.pdf

**Comment by jaustermann** [comment]
I think this is a worthwhile addition. The two values that are generally used for density_above are 0 (~air), which is what we use as a default, and 1000 (~water). People use both and sometimes a combination (0 above continents and 1000 above oceans). This density contrast can also be calculated self-consistently for a viscoelastic Earth (cough ... http://gji.oxfordjournals.org/content/203/3/1909.full).

I don't have write permission to comment directly on the commit, but here is my suggestion to maybe reword the description to make the parameter a little clearer ... just a suggestion ...

"Dynamic topography is calculated as the excess (or lack of) mass that is supported by mantle flow. This value depends on the density of material that is moved up (or down, i.e. crustal rock) and the density of the material that is displaced (generally water or air). While the density of crustal rock is part of the material model, this parameter 'density_above' allows the user to specify the density value of material that is displaced above the solid surface. By default this material is assumed to be air (with a density of 0 kg/m3)."

**Comment by Shangxin-Liu** [comment]
Thanks for the detailed comment, Jacky. Yes. Your description is more 'geophysical'. I can change the description to yours but let's leave the final decision to Wolfgang and Rene.

**Comment by Shangxin-Liu** [comment]
Perhaps we should also enable the combination calculation between 0 above continent and 1020 above oceans. I already have a 'diff ocean and continent location' patch but haven't put it to utilities.cc file. I will do it later.

**Comment by bangerth** [comment]
@jaustermann -- you have the right to add comments to individual patches. I think I added you to the list of people who can do so a good long while ago :-)

@Shangxin-Liu -- I'll gladly defer to Jacky regarding the exact wording. If she suggests particular wording, I will support this suggestion.

**Comment by Shangxin-Liu** [comment]
Ok. This is the parameter description from Jacky's comments.

File: source/postprocess/dynamic_topography.cc [hunk 2]
```diff
@@ -257,7 +257,15 @@ namespace aspect
                              "in the outputted data file (not visualization). "
                              "'true' subtracts the mean, 'false' leaves "
                              "the calculated dynamic topography as is. ");
-
+          prm.declare_entry ("Density above","0",
+                             Patterns::Double (0),
+                             "Dynamic topography is calculated as the excess or lack of mass that is supported by mantle flow. "
+                             "This value depends on the density of material that is moved up or down, i.e. crustal rock, and the "
+                             "density of the material that is displaced (generally water or air). While the density of crustal rock "
+                             "is part of the material model, this parameter 'density_above' allows the user to specify the density "
```
Comment (gassmoeller) [inline source/postprocess/dynamic_topography.cc]: Please replace the name of the variable by the name of the parameter, i.e. change 'density_above' to 'Density above' in the description.

File: source/postprocess/visualization/dynamic_topography.cc [hunk 3]
```diff
@@ -207,7 +207,15 @@ namespace aspect
                                  "in the outputted data file (not visualization). "
                                  "'true' subtracts the mean, 'false' leaves "
                                  "the calculated dynamic topography as is. ");
-
+              prm.declare_entry ("Density above","0",
+                                 Patterns::Double (0),
+                                 "Dynamic topography is calculated as the excess or lack of mass that is supported by mantle flow. "
+                                 "This value depends on the density of material that is moved up or down, i.e. crustal rock, and the "
+                                 "density of the material that is displaced (generally water or air). While the density of crustal rock "
+                                 "is part of the material model, this parameter 'density_above' allows the user to specify the density "
```
Comment (gassmoeller) [inline source/postprocess/visualization/dynamic_topography.cc]: Same as above.

**Comment by gassmoeller** [comment]
Looks good. After addressing my minor comments this is ready to merge.

File: source/postprocess/visualization/dynamic_topography.cc (hunk 3 above)
Comment (Shangxin-Liu) [inline source/postprocess/visualization/dynamic_topography.cc]: Just fixed this minor comment.

Comment (gassmoeller) [inline source/postprocess/visualization/dynamic_topography.cc]: Hi Shangxin,
last minor thing: I meant to replace 'density_above' by 'Density above' (the name of the input parameter, with space instead of underscore) so that users can directly see what to change in the input file. Maybe the formatting was lost on Github.

Comment (Shangxin-Liu) [inline source/postprocess/visualization/dynamic_topography.cc]: Oh, that's my negligence.. Just fixed this:)

Comment (bangerth) [inline source/postprocess/visualization/dynamic_topography.cc]: While you're there: the text you enter here is used as the input of LaTeX. So it would be rendered more correctly if you wrote this as
```
  ...`Density above'...
```
i.e., with a backtick at the front and a single regular tick at the end of the text.
